optuna/pruners/threshold.py

```diff
@@ -0,0 +1,116 @@
+import math
+from typing import Optional
+
+from optuna.pruners import BasePruner
+from optuna.pruners.percentile import _is_first_in_interval_step
+
+import optuna
+
+
+class ThresholdPruner(BasePruner):
+    """Pruner to detect outlying metrics of the trials.
+
+    Prune if a metric exceeds upper bound threshold or
+    falls behind lower bound threshold.
+
+    Example:
+        .. testcode::
+
+            from optuna import create_study
+            from optuna.exceptions import TrialPruned
+            from optuna.pruners import ThresholdPruner
+
+
+            def objective_1(trial):
+                for step in range(n_trial_step):
+                    trial.report(ys_for_upper[step], step)
+
+                    if trial.should_prune():
+                        raise TrialPruned()
+
+
+            def objective_2(trial):
+                for step in range(n_trial_step):
+                    trial.report(ys_for_lower[step], step)
+
+                    if trial.should_prune():
+                        raise TrialPruned()
+
+
+            ys_for_upper = [0.0, 0.1, 0.2, 0.5, 1.2]
+            ys_for_lower = [100.0, 90.0, 0.1, 0.0, -1]
+            n_trial_step = 5
+
+            study = create_study(pruner=ThresholdPruner(upper=1.0))
+            study.optimize(objective_1, n_trials=10)
+
+            study = create_study(pruner=ThresholdPruner(lower=0.0))
+            study.optimize(objective_2, n_trials=10)
+
+
+    Args
+        lower:
+            minimum value which determines whether pruner prunes or not
+            (If value is smaller than lower, it prunes)
+        upper:
+            maximum value which determines whether pruner prunes or not
+            (If value is larger than upper, it prunes)
+        n_startup_trials:
+            Pruning is disabled until the given number of trials finish in the same study.
+        n_warmup_steps:
+            Pruning is disabled until the trial exceeds the given number of step.
+        interval_steps:
+            Interval in number of steps between the pruning checks, offset by the warmup steps.
+            If no value has been reported at the time of a pruning check, that particular check
+            will be postponed until a value is reported. Value must be at least 1.
+
+    """
+
+    def __init__(
+            self,
+            lower: Optional[float] = None,
+            upper: Optional[float] = None,
+            n_warmup_steps: int = 0,
+            interval_steps: int = 1
+    ) -> None:
+
+        if isinstance(lower, float) and isinstance(upper, float) and lower > upper:
+            raise ValueError('lower should be smaller than upper')
```

Comment:
I think you must not check the types of lower and upper.

```console
$ python
Python 3.8.1 (default, Jan 17 2020, 14:59:07) 
[Clang 11.0.0 (clang-1100.0.33.17)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> def foo(lower: float, upper: float):
...     if isinstance(lower, float) and isinstance(upper, float) and lower > upper:
...         raise ValueError('lower should be smaller than upper')
... 
>>> foo(3, 2)
>>>
```

How about the following?

```suggestion
        if lower is not None and upper is not None and lower > upper:
            raise ValueError('lower should be smaller than upper')
```

Reply:
Thank you for catching this! I missed it.